fix(compiler-cli): fix extending angularCompilerOptions from non relative extension less TypeScript configuration files

JounQin opened the pull request:
close #41343

cc @alan-agius4 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #41343


## What is the new behavior?

support sub file of node package and relative path without json extension

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

alan-agius4 commented:
Hi @JounQin,

Thanks for this, I’ll try to review this next week as had some days off.

Just 2 points.

1. The CI failure seems to be related to this change
2. When you check for extension, let’s validate that these are json.
2. Can try to implant this change without using TSC private API? I don’t think it should be too difficult to achieve the same behaviour using the public API.

alan-agius4 commented:
Also maybe the commit message can be changed to provide additional context.

```
fix(compiler-cli): fix extending angularCompilerOptions from unrooted and extension less TypeScript configurations TypeScript

JounQin commented:
> Hi @JounQin,
> 
> Thanks for this, I’ll try to review this next week as had some days off.
> 
> Just 3 points to keep in mind.
> 
> 1. The CI failure seems to be related to this change
> 2. When you check for extension, let’s validate that these are json.
> 3. Can try to implement this change without using TSC private API? I don’t think it should be too difficult to achieve the same behaviour using the public API. We should avoid using private APIs.

@alan-agius4 

1. I have no idea why the CI failed. `Error: ENOENT: no such file or directory, unlink 'C:\users\circleci\_bazel_circleci\u4uoan2j\execroot\angular\node_modules\@angular\compiler'`, I don't have a Windows PC personally, do you have any idea to fix?

fixed 2 and 3

alan-agius4 commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -249,6 +251,19 @@ function getExtendedConfigPath(
             parseConfigHost);
     if (resolvedModule) {
       extendedConfigPath = absoluteFrom(resolvedModule.resolvedFileName);
+    } else
```
If we use a recursive function this code can be simplified, also for consistence either use  `host.extname` or `.endsWith`

```ts
function getExtendedConfigPath(
  configFile: string,
  extendsValue: string,
  host: ConfigurationHost,
  fs: FileSystem,
  forceJsonExtension = false,
): AbsoluteFsPath | null {
  let extendedConfigPath: AbsoluteFsPath | null = null;

  if (forceJsonExtension && !extendsValue.endsWith('.json')) {
    extendsValue = `${extendsValue}.json`;
  }

  if (extendsValue.startsWith('.') || fs.isRooted(extendsValue)) {
    extendedConfigPath = host.resolve(host.dirname(configFile), extendsValue);
  } else {
    const parseConfigHost = createParseConfigHost(host, fs);

    // Path isn't a rooted or relative path, resolve like a module.
    const {
      resolvedModule,
    } = ts.nodeModuleNameResolver(
      extendsValue, configFile,
      { moduleResolution: ts.ModuleResolutionKind.NodeJs, resolveJsonModule: true },
      parseConfigHost);
    if (resolvedModule) {
      extendedConfigPath = absoluteFrom(resolvedModule.resolvedFileName);
    }
  }

  if (host.exists(extendedConfigPath) || extendedConfigPath.endsWith('.json') || forceJsonExtension) {
    return extendedConfigPath;
  }

  return getExtendedConfigPath(configFile, extendsValue, host, fs, true);
}
```

alan-agius4 commented on packages/compiler-cli/test/perform_compile_spec.ts:
```diff
@@ -136,4 +136,62 @@ describe('perform_compile', () => {
     }));
   });
+
+  it('should merge tsconfig "angularCompilerOptions" when extends point to sub file of node package',
```
```suggestion
  it('should merge tsconfig "angularCompilerOptions" when extends point to file without extension in a node package',
```

alan-agius4 reviewed (changes requested):
Thanks for this.

2 small nits.

alan-agius4 commented:
Can you also please add `close #41343` as commit message footer.

AndrewKushnir commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -233,7 +233,9 @@ function getExtendedConfigPath(
   if (extendsValue.startsWith('.') || fs.isRooted(extendsValue)) {
     extendedConfigPath = host.resolve(host.dirname(configFile), extendsValue);
-    extendedConfigPath = host.extname(extendedConfigPath) ?
+    // https://github.com/microsoft/TypeScript/blob/4774666f2343c6ffdbf862ffff63f8653a09449b/src/compiler/commandLineParser.ts#L2806
```
Quick comment: could you please add a couple sentences in the code (as inline comments) describing this, so that there is always a quick context available.

JounQin committed:
fix(compiler-cli): fix extending angularCompilerOptions from non relative extension less TypeScript configuration files

support non rooted file of node package and relative path without json extension

close #41343

JounQin commented:
@alan-agius4 I have no idea why CI is broken again, and it seems the master branch is also broken. (I've just rebased from the master branch)

alan-agius4 reviewed (approved):
LGTM, thanks

alan-agius4 commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -227,15 +227,23 @@ function createParseConfigHost(host: ConfigurationHost, fs = getFileSystem()): t
+    ): AbsoluteFsPath|null {
   let extendedConfigPath: AbsoluteFsPath|null = null;
 
+  // resolve `.json` module file automatically like TypeScript
+  // @1stg/tsconfig/angular -> @1stg/tsconfig/angular.json
+  // ./tsconfig.app -> ./tsconfig.app.json
```
Nit 
```suggestion
```

In think it's clear what this is doing, I think it's better to add the comment when we use `forceJsonExtension`  to explain why.

alan-agius4 commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -252,12 +260,15 @@ function getExtendedConfigPath(
   }
 
-  return null;
+  return getExtendedConfigPath(configFile, extendsValue, host, fs, true);
```
Nit:
 
```suggestion
  // Try to resolve the paths with a json extension append a json extension to the file in case if is is missing and the resolution failed. This is to replicate TypeScript behaviour
  // See: https://github.com/microsoft/TypeScript/blob/4774666f2343c6ffdbf862ffff63f8653a09449b/src/compiler/commandLineParser.ts#L2806
  return getExtendedConfigPath(configFile,`{extendsValue}.json`, host, fs);
```

alan-agius4 commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -227,15 +227,23 @@ function createParseConfigHost(host: ConfigurationHost, fs = getFileSystem()): t
+    extendsValue: string,
+    host: ConfigurationHost,
+    fs: FileSystem,
+    forceJsonExtension = false,
```
```suggestion
    forceJsonExtension = false,
```

alan-agius4 commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -227,15 +227,23 @@ function createParseConfigHost(host: ConfigurationHost, fs = getFileSystem()): t
+  // resolve `.json` module file automatically like TypeScript
+  // @1stg/tsconfig/angular -> @1stg/tsconfig/angular.json
+  // ./tsconfig.app -> ./tsconfig.app.json
+  if (forceJsonExtension && !extendsValue.endsWith('.json')) {
```
```suggestion
```

alan-agius4 commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -227,15 +227,23 @@ function createParseConfigHost(host: ConfigurationHost, fs = getFileSystem()): t
+    ): AbsoluteFsPath|null {
   let extendedConfigPath: AbsoluteFsPath|null = null;
 
+  // resolve `.json` module file automatically like TypeScript
+  // @1stg/tsconfig/angular -> @1stg/tsconfig/angular.json
+  // ./tsconfig.app -> ./tsconfig.app.json
+  if (forceJsonExtension && !extendsValue.endsWith('.json')) {
+    extendsValue = `${extendsValue}.json`;
+  }
+
```
Actually I just figured that this is not needed at all. Sorry about this.

```suggestion
```

alan-agius4 reviewed (changes requested):
Actually, just realised that `forceJsonExtension` is not needed and we can simplify the method even more. Sorry about this.

JounQin commented:
> Actually, just realised that `forceJsonExtension` is not needed and we can simplify the method even more. Sorry about this.

Then we need to change `extendedConfigPath.endsWith('.json')` to `extendsValue.endsWith('.json')` right?

alan-agius4 commented:
Yes

alan-agius4 commented:
Yes and for performance reasons the check should be inverted
```ts
if (extendsValue?.endsWith('.json') || host.exists(extendedConfigPath))** {
```

JounQin commented:
@alan-agius4 Done!

alan-agius4 reviewed (approved):
LGTM, thanks

JounQin commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -252,12 +252,17 @@ function getExtendedConfigPath(
 
-  return null;
+  // Try to resolve the paths with a json extension append a json extension to the file in case if
+  // is is missing and the resolution failed. This is to replicate TypeScript behaviour See:
```
@alan-agius4 Should it be `it is`? I’ve just copied your comments.

alan-agius4 replied:
Yeah absolutely

JoostK commented on packages/compiler-cli/test/perform_compile_spec.ts:
```diff
@@ -136,4 +136,62 @@ describe('perform_compile', () => {
     }));
   });
+
+  it('should merge tsconfig "angularCompilerOptions" when extends point to an extension less non rooted file',
```
How is this test different from the existing one above? It looks like the same scenario to me.

alan-agius4 replied:
That test is different because the extends point to a package and the path of the tsconfig is resolved from https://github.com/angular/angular/blob/07fe4dc6267a25fdea45cd2f29c705a82add9663/packages/compiler-cli/test/perform_compile_spec.ts#L127

In this case the extends point to an extension less json file in a package, which resolution is slightly different, as module resolution will fail in this case and needs to be invoked with a json extension. 

Ie: in this case module resolution will be called twice, once without extension and when that fails it will be called again with the json extension.

JoostK commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -252,12 +252,17 @@ function getExtendedConfigPath(
+  // Try to resolve the paths with a json extension append a json extension to the file in case if
+  // it is missing and the resolution failed. This is to replicate TypeScript behaviour See:
+  // https://github.com/microsoft/TypeScript/blob/294a5a7d784a5a95a8048ee990400979a6bc3a1c/src/compiler/commandLineParser.ts#L2806
+  return getExtendedConfigPath(configFile, `${extendsValue}.json`, host, fs);
```
I think I prefer the way TypeScript structured this code, instead of the recursive invocation. The recursion makes it less obvious that only two files are attempted, and the recursion is actually not relevant for the module resolution case (because the result from module resolution has already been verified to exist) but that's not clear from the code paths.

I would suggest to make the change closer to TypeScript's logic. The `if (extendedConfigPath !== null && host.exists(extendedConfigPath))` condition that used to be at the top-level of this function can actually be moved into the block for relative and rooted files, which is where the json file handling can then be handled as well. You can then even remove the separate declaration of the `extendedConfigPath` variable as it won't be necessary in the module resolution block anymore; that one can be simplified to `return absoluteFrom(resolvedModule.resolvedFileName);`. This results in the exact same structure as TS uses.

alan-agius4 replied:
Note: module resolution might fail in which case it needs to be invoked again with the json extension. 

TS logic doesn’t need this because they have an option `lookupConfig ` which is not exposed as part of the API that allows this behaviour. 

https://github.com/microsoft/TypeScript/blob/b34e705672cc31b86ce9291f1e234b387aaa3184/src/compiler/moduleNameResolver.ts#L925

JoostK replied:
Ooh, sneaky. I still think that the overarching `extendedConfigPath` variable is redundant, though, as its assignment in the module resolution block can be replaced with an immediate return. The recursion can be avoided by having a separate function that does the work, which can then be invoked twice:

```ts
function getExtendedConfigPath(configFile: string, extendsValue: string, host: ConfigurationHost, fs: FileSystem): AbsoluteFsPath|null {
  const result = getExtendedConfigPathWorker(configFile, extendsValue, host, fs);
  if (result !== null) {
    return result;
  }

  // Try to resolve the paths with a json extension append a json extension to the file in case if
  // it is missing and the resolution failed. This is to replicate TypeScript behaviour, see:
  // https://github.com/microsoft/TypeScript/blob/294a5a7d784a5a95a8048ee990400979a6bc3a1c/src/compiler/commandLineParser.ts#L2806
  return getExtendedConfigPathWorker(configFile, `${extendsValue}.json`, host, fs);
}

function getExtendedConfigPathWorker(configFile: string, extendsValue: string, host: ConfigurationHost, fs: FileSystem): AbsoluteFsPath|null {
  if (extendsValue.startsWith('.') || fs.isRooted(extendsValue)) {
    const extendedConfigPath = host.resolve(host.dirname(configFile), extendsValue);
    if (host.exists(extendedConfigPath))) {
      return extendedConfigPath;
    }
  } else {
    const parseConfigHost = createParseConfigHost(host, fs);

    // Path isn't a rooted or relative path, resolve like a module.
    const {
      resolvedModule,
    } =
        ts.nodeModuleNameResolver(
            extendsValue, configFile,
            {moduleResolution: ts.ModuleResolutionKind.NodeJs, resolveJsonModule: true},
            parseConfigHost);
    if (resolvedModule) {
      return absoluteFrom(resolvedModule.resolvedFileName);
    }
  }

  return null;
}
```

I don't feel too strongly about this, though, so happy to approve the current approach.

JounQin replied:
@alan-agius4 What's your opinion here? I'm fine with this change or not.

alan-agius4 replied:
I am good either way. But let's go with @JoostK approach since it's a bit easier to follow.

JounQin replied:
Done!

JoostK commented on packages/compiler-cli/test/perform_compile_spec.ts:
```diff
@@ -136,4 +136,62 @@ describe('perform_compile', () => {
     }));
   });
+
+  it('should merge tsconfig "angularCompilerOptions" when extends point to an extension less non rooted file',
```
```suggestion
  it('should merge tsconfig "angularCompilerOptions" when extends points to a non rooted file without json extension',
```

JoostK commented on packages/compiler-cli/test/perform_compile_spec.ts:
```diff
@@ -136,4 +136,62 @@ describe('perform_compile', () => {
+       }));
+     });
+
+  it('should merge tsconfig "angularCompilerOptions" when extends point to relative config without json extension',
```
```suggestion
  it('should merge tsconfig "angularCompilerOptions" when extends points to relative config without json extension',
```

JoostK reviewed (approved):
Thank you for working on this. Some nits, but happy to approve.

AndrewKushnir commented:
[Presubmit](http://test/OCL:367049056:BASE:367049088:1617734035013:27be8d22).
